File: chessboard_and_queens/chessboard_and_queens.py
from sys import stdin
from pprint import pprint

def read_multiple(from_str_constructor, delimiter=''):
    return list(map(from_str_constructor, stdin.readline().strip().split(sep=delimiter)))

def read_single(from_str_constructor):
    return from_str_constructor(stdin.readline().strip())

def read_matrix(num_rows , str_to_row_constructor):
    l = []
    for _ in range(num_rows):
        l.append(str_to_row_constructor(stdin.readline().strip()))
    return l

def main():
    N = 8
    board = read_matrix(N, lambda s: list(map(lambda x: True if x == '.' else False, s)))
    used_positions = set()
    num_ways = 0
    def can_use(row, col):
        if not board[row][col]:
            return False
        for used_row, used_col in used_positions:
            if row == used_row or col == used_col:
                return False
            if abs(row - used_row) == abs(col - used_col):
                return False
        return True
    def dfs(row=0):
        nonlocal num_ways
        if row >= N:
            return
        for col in range(N):
            if can_use(row, col):
                if row == N - 1:
                    num_ways += 1
                used_positions.add((row,col))
                dfs(row+1)
                used_positions.remove((row,col))

    dfs()
    print(num_ways)

main()

File: chessboard_and_queens/chessboard_and_queens.cc
#include <iostream>
#include <vector>
#include <set>
#include <array>
#include <math.h>

using namespace std;

template <class T>
void read_many(int n, vector<T>& v){
  for (int i = 0; i < n; ++i) {
    T tmp;
    cin >> tmp;
    v.push_back(tmp);
  }
}

template <class T>
void read_matrix(int m, int n, vector <vector <T> >& v) {
  for (int i = 0; i < m; ++i) {
    vector<T> vi;
    read_many(n, vi);
    v.push_back(vi);
  }
}

char v[8][8];
set<array<int, 2> > used;
int numWays;

bool can_use(int row, int col){
  if (v[row][col] == '*') {
    return false;
  }
  for (const auto& [r, c]: used) {
    if (r == row || c == col) {
      return false;
    } else if (abs(row - r) == abs(col - c)) {
      return false;
    }
  }
  return true;
}

void dfs(int row, const int n){
  if (row >= n) {
    return;
  } 
  for (int i = 0; i < n; ++i) {
    if (can_use(row, i)) {
      if (row == n-1) {
        numWays +=1;
      }
      used.insert(array{row, i});
      dfs(row+1, n);
      used.erase(array{row, i});
    }
  }
}

int main() {
  string in;
  for (int i = 0; i < 8; ++i){
    cin >> in;
    for (int j = 0; j < 8; ++j) {
      v[i][j] = in[j];
    }
  }
  dfs(0, 8);
  cout << numWays << endl;
}
